### DISCUSSION TITLE
Taking a screenshot in avalonia 11

### DISCUSSION THREAD

**Discussion by Varmod** [AvaloniaUI/Avalonia#11669] [Q&A] [answered]
In my app I need the ability to take a screenshot, but I am unable figure out way how to do it.
I've found an older thread where it was recommended to use the the code from the TestBase class 
https://github.com/AvaloniaUI/Avalonia/blob/release/11.0.0-rc1.1/tests/Avalonia.RenderTests/TestBase.cs
but I wasn't able to get that code running as many of the Compositor methods are internal.

Is there some other way to do it? or perhaps a way to directly access the last rendered frame?

**Comment by maxkatz6** [comment, marked as answer]
The only important part is RenderTargetBitmap, which is still public.
```csharp
            var pixelSize = new PixelSize((int)target.Width, (int)target.Height);
            var size = new Size(target.Width, target.Height);
            var dpiVector = new Vector(dpi, dpi);

            using (RenderTargetBitmap bitmap = new RenderTargetBitmap(pixelSize, dpiVector))
            {
                target.Measure(size);
                target.Arrange(new Rect(size));
                bitmap.Render(target);
                bitmap.Save(immediatePath);
            }
```

>  or perhaps a way to directly access the last rendered frame?

No.

**Reply by Varmod** [reply]
When using this approach if I have a combo-box or menu open it won't get rendered onto to the bitmap. What would be the correct approach to also get these incorporated onto the bitmap?

I'm currently rendering the main window.

**Reply by timunie** [reply]
try to enable OverlayPopups (defined in AppBuilder). This may help in this situation. You are right, normal Popups have their own window and thus are not included in the image.

**Reply by Varmod** [reply]
Thank you for your quick reply, this solved the issue for me.

**Reply by ISDHN** [reply]
Could you please tell me where the *target* come from? How can I get the *target* object?

**Reply by rabbitism** [reply]
> Could you please tell me where the *target* come from? How can I get the *target* object?

It is the control you want to render.

**Reply by ISDHN** [reply]
So if I want to get the screenshot of the whole screen, what should i pass?

**Reply by rabbitism** [reply]
> So if I want to get the screenshot of the whole screen, what should i pass? 

Then you need to interop with your os. It's not related to Avalonia.

**Comment by BYJRK** [comment]
I have two more questions:

1. How is the dpi parameter calculated? After testing, I found that there seems to be a default value of 96, and then the `pixelSize` needs to be converted proportionally. Is this best practice? 
2. I noticed that if a control includes `DropShadowEffect`, this effect will be ignored. How can I retain the shadows?

**Reply by Coloryr** [reply]
DPI默认就行
DPI use default is well
Shadow需要叠个白底才看得见，然后需要用margin撑开才能看得见
Shadow need a white BG, and set margin
